perun/profile/convert.py

```diff
@@ -143,15 +147,27 @@ def to_flame_graph_format(
             if "subtype" not in alloc.keys() or alloc["subtype"] != "free":
                 # Workaround for total time used in some collectors, so it is not outputted
                 if alloc["uid"] != "%TOTAL_TIME%" and profile_key in alloc:
-                    stack_str = to_uid(alloc["uid"]) + ";"
-                    for frame in alloc.get("trace", [])[::-1]:
+                    stack = alloc.get("trace", [])
+                    stack.append(alloc["uid"])
+                    stack_str = ""
+                    unknown_cnt = 0
+                    for frame in reversed(stack):
                         line = to_uid(frame, minimize)
-                        stack_str += line + ";"
-                    if stack_str and stack_str.endswith(";"):
-                        final = stack_str[:-1]
-                        final += " " + str(alloc[profile_key]) + "\n"
-                        stacks.append(final)
-
+                        if squash_unknown:
+                            # Fold unknown towers
+                            if line == "[unknown]":
+                                unknown_cnt += 1
+                                continue
+                            elif unknown_cnt > 0:
+                                stack_str += "[unknown[squashed]];"
+                                unknown_cnt = 0
+                        stack_str += f"{line};"
+                    # The unknown frame might have been the last one, in which case the frame was
+                    # not yet added to the string.
+                    if unknown_cnt > 0:
+                        stack_str += "[unknown[squashed]];"
```

Comment:
Would it make sense to output this as `[unknown[squashed]]^{unknown_cnt}` or something similar to know how much you squashed?

Reply:
Ideally, yes, that was the original intention. However, it (a) may break a single parent unknown frame into multiple smaller frames, which changes the graph topology in a bit unintuitive way, and (b) it complicates the mouseover highlighting, as it relies on exact match. As stated in the PR message, this is more of an ad-hoc fix that I plan to redo later as part of refactoring the data handling in report.

Reply:
I thought it squashed all unknows in tower to single unit. Can you maybe show a picture/examples how this squashes the towers?

Reply:
The current solution does exactly that, but the proposed solution with `[unknown[squashed]]^{unknown_cnt}` would not, as the towers are in fact often trees (i.e., they branch into multiple branches all having a different leaf function). Hence, the tree would be split into strictly linear traces, all with different names (e.g., `[unknown[squashed]]^2`, `[unknown[squashed]]^5`, etc.) and the flame graph algorithm would not be able to merge them that way.